### DISCUSSION TITLE
mill and JAVA_HOME under ubuntu

### DISCUSSION THREAD

**Discussion by nemoo** [com-lihaoyi/mill#1818] [Q&A] [answered]
I am having problems on ubuntu with mill finding the java home.
I tried this:

    > export JAVA_HOME=/usr/bin/java                          
    > mill concat
    /home/username/.cache/mill/download/0.10.2: 43: exec: /usr/bin/java/bin/java: not found

mill seems to add "/bin/java" to my JAVA_HOME variable?

If I make the JAVA_HOME path shorter accordingly, mill seems to find java and runs fine:

    > export JAVA_HOME=/usr         
    > mill concat          
    [1/2] srcs 

Does that make any sense?

**Comment by lefou** [comment, marked as answer]
As `JAVA_HOME` is just a convention but not a standard, most (if not all) tools that support this variable expect a JDK/JRE directory layout. That's why we add (as you already found out) a `bin/java` to access the Java executable. If your Java installation is rolled out into a LSB directory layout, using `JAVA_HOME=/usr` makes sense. But I can't tell you if there are any downsides.
